File: android/src/main/java/com/example/flutter_naver_map/FlutterNaverMapPlugin.java
package com.example.flutter_naver_map;

import android.app.Activity;
import android.app.Application;
import android.os.Bundle;

import androidx.annotation.NonNull;
import androidx.lifecycle.DefaultLifecycleObserver;
import androidx.lifecycle.Lifecycle;
import androidx.lifecycle.LifecycleOwner;

import java.util.concurrent.atomic.AtomicInteger;
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding;
import io.flutter.embedding.engine.plugins.FlutterPlugin;
import io.flutter.embedding.engine.plugins.activity.ActivityAware;
import io.flutter.embedding.engine.plugins.lifecycle.FlutterLifecycleAdapter;
import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;
import io.flutter.plugin.common.MethodChannel.MethodCallHandler;
import io.flutter.plugin.common.MethodChannel.Result;
import io.flutter.plugin.common.PluginRegistry.Registrar;

/** FlutterNaverMapPlugin */
public class FlutterNaverMapPlugin implements Application.ActivityLifecycleCallbacks, FlutterPlugin, ActivityAware, DefaultLifecycleObserver {
  static final int CREATED = 1;
  static final int STARTED = 2;
  static final int RESUMED = 3;
  static final int PAUSED = 4;
  static final int STOPPED = 5;
  static final int DESTROYED = 6;
  private final AtomicInteger state = new AtomicInteger(0);
  private int registrarActivityHashCode;
  private FlutterPluginBinding pluginBinding;
  private static final String VIEW_TYPE = "flutter_naver_map";
  private Lifecycle lifecycle;

  @Override
  public void onAttachedToEngine(@NonNull FlutterPluginBinding flutterPluginBinding) {
//    final MethodChannel channel = new MethodChannel(flutterPluginBinding.getFlutterEngine().getDartExecutor(), "flutter_naver_map");
//    channel.setMethodCallHandler(new FlutterNaverMapPlugin());
    pluginBinding = flutterPluginBinding;
  }

  // This static function is optional and equivalent to onAttachedToEngine. It supports the old
  // pre-Flutter-1.12 Android projects. You are encouraged to continue supporting
  // plugin registration via this function while apps migrate to use the new Android APIs
  // post-flutter-1.12 via https://flutter.dev/go/android-project-migration.
  //
  // It is encouraged to share logic between onAttachedToEngine and registerWith to keep
  // them functionally equivalent. Only one of onAttachedToEngine or registerWith will be called
  // depending on the user's project. onAttachedToEngine or registerWith must both be defined
  // in the same class.
  public static void registerWith(Registrar registrar) {
//    final MethodChannel channel = new MethodChannel(registrar.messenger(), "flutter_naver_map");
//    channel.setMethodCallHandler(new FlutterNaverMapPlugin());
    if (registrar.activity() == null) {
      // When a background flutter view tries to register the plugin, the registrar has no activity.
      // We stop the registration process as this plugin is foreground only.
      return;
    }
    final FlutterNaverMapPlugin plugin = new FlutterNaverMapPlugin(registrar.activity());
    registrar.activity().getApplication().registerActivityLifecycleCallbacks(plugin);
    registrar
            .platformViewRegistry()
            .registerViewFactory(
                    VIEW_TYPE,
                    new NaverMapFactory(plugin.state, registrar.messenger(), null, null, registrar, -1));
  }

  @Override
  public void onDetachedFromEngine(@NonNull FlutterPluginBinding binding) {
    pluginBinding = null;
  }

  @java.lang.Override
  public void onAttachedToActivity(ActivityPluginBinding activityPluginBinding) {
    lifecycle = FlutterLifecycleAdapter.getActivityLifecycle(activityPluginBinding);
    lifecycle.addObserver(this);
    pluginBinding
            .getPlatformViewRegistry()
            .registerViewFactory(
                    VIEW_TYPE,
                    new NaverMapFactory(
                            state,
                            pluginBinding.getBinaryMessenger(),
                            activityPluginBinding.getActivity().getApplication(),
                            lifecycle,
                            null,
                            activityPluginBinding.getActivity().hashCode()));

  }

  @java.lang.Override
  public void onDetachedFromActivityForConfigChanges() {
    this.onDetachedFromActivity();
  }

  @java.lang.Override
  public void onReattachedToActivityForConfigChanges(ActivityPluginBinding activityPluginBinding) {
    lifecycle = FlutterLifecycleAdapter.getActivityLifecycle(activityPluginBinding);
    lifecycle.addObserver(this);
  }

  @java.lang.Override
  public void onDetachedFromActivity() {
    lifecycle.removeObserver(this);
  }


  // DefaultLifecycleObserver methods

  @Override
  public void onCreate(@NonNull LifecycleOwner owner) {
    state.set(CREATED);
  }

  @Override
  public void onStart(@NonNull LifecycleOwner owner) {
    state.set(STARTED);
  }

  @Override
  public void onResume(@NonNull LifecycleOwner owner) {
    state.set(RESUMED);
  }

  @Override
  public void onPause(@NonNull LifecycleOwner owner) {
    state.set(PAUSED);
  }

  @Override
  public void onStop(@NonNull LifecycleOwner owner) {
    state.set(STOPPED);
  }

  @Override
  public void onDestroy(@NonNull LifecycleOwner owner) {
    state.set(DESTROYED);
  }


  @Override
  public void onActivityCreated(Activity activity, Bundle bundle) {
    if (activity.hashCode() != registrarActivityHashCode) {
      return;
    }
    state.set(CREATED);
  }

  @Override
  public void onActivityStarted(Activity activity) {
    if (activity.hashCode() != registrarActivityHashCode) {
      return;
    }
    state.set(STARTED);
  }

  @Override
  public void onActivityResumed(Activity activity) {
    if (activity.hashCode() != registrarActivityHashCode) {
      return;
    }
    state.set(RESUMED);
  }

  @Override
  public void onActivityPaused(Activity activity) {
    if (activity.hashCode() != registrarActivityHashCode) {
      return;
    }
    state.set(PAUSED);
  }

  @Override
  public void onActivityStopped(Activity activity) {
    if (activity.hashCode() != registrarActivityHashCode) {
      return;
    }
    state.set(STOPPED);
  }

  @Override
  public void onActivitySaveInstanceState(Activity activity, Bundle bundle) {

  }

  @Override
  public void onActivityDestroyed(Activity activity) {
    if (activity.hashCode() != registrarActivityHashCode) {
      return;
    }
    activity.getApplication().unregisterActivityLifecycleCallbacks(this);
    state.set(DESTROYED);
  }

  public FlutterNaverMapPlugin() {}

  private FlutterNaverMapPlugin(Activity activity) {
    this.registrarActivityHashCode = activity.hashCode();
  }
}

File: android/src/main/kotlin/com/example/flutter_naver_map/FlutterNaverMapPlugin.kt
package com.example.flutter_naver_map

import android.app.Activity
import android.app.Application
import android.os.Bundle
import androidx.annotation.NonNull;
import androidx.lifecycle.DefaultLifecycleObserver
import androidx.lifecycle.Lifecycle
import androidx.lifecycle.LifecycleOwner
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import io.flutter.plugin.common.PluginRegistry.Registrar
import java.util.concurrent.atomic.AtomicInteger

/** FlutterNaverMapPlugin */
public class FlutterNaverMapPlugin: Application.ActivityLifecycleCallbacks, FlutterPlugin, ActivityAware, DefaultLifecycleObserver {
  private val registrarActivityHashCode: Int = 0
  private val state = AtomicInteger(0)
  internal val CREATED = 1
  internal val STARTED = 2
  internal val RESUMED = 3
  internal val PAUSED = 4
  internal val STOPPED = 5
  internal val DESTROYED = 6
  private var pluginBinding: FlutterPlugin.FlutterPluginBinding? = null

  override fun onActivityPaused(activity: Activity?) {
    if (activity.hashCode() != registrarActivityHashCode) {
      return
    }
    state.set(PAUSED)
  }

  override fun onActivityResumed(activity: Activity?) {
    if (activity.hashCode() != registrarActivityHashCode) {
      return
    }
    state.set(RESUMED)
  }

  override fun onActivityStarted(activity: Activity?) {
    if (activity.hashCode() != registrarActivityHashCode) {
      return
    }
    state.set(STARTED)
  }

  override fun onActivityDestroyed(activity: Activity?) {
    if (activity.hashCode() != registrarActivityHashCode) {
      return
    }
    activity?.getApplication()?.unregisterActivityLifecycleCallbacks(this)
    state.set(DESTROYED)
  }

  override fun onActivitySaveInstanceState(p0: Activity?, p1: Bundle?) {
  }

  override fun onActivityStopped(activity: Activity?) {
    if (activity.hashCode() != registrarActivityHashCode) {
      return
    }
    state.set(STOPPED)
  }

  override fun onActivityCreated(activity: Activity?, p1: Bundle?) {
    if (activity.hashCode() != registrarActivityHashCode) {
      return
    }
    state.set(CREATED)
  }

  override fun onDetachedFromActivity() {
  }

  override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) {
  }

  override fun onAttachedToActivity(binding: ActivityPluginBinding) {
//    binding
//            .getPlatformViewsController()
//            .getRegistry()
//            .registerViewFactory(
//                    "flutter_naver_map", FlutterNaverMapViewFactory(messenger, /*containerView=*/ null))
    pluginBinding
            ?.getPlatformViewRegistry()
            ?.registerViewFactory(
                    "flutter_naver_map",
                    FlutterNaverMapViewFactory(pluginBinding!!.binaryMessenger, /*containerView=*/ null))
  }

  override fun onDetachedFromActivityForConfigChanges() {
    TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
  }


  override fun onAttachedToEngine(@NonNull flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
//    val channel = MethodChannel(flutterPluginBinding.getFlutterEngine().getDartExecutor(), "flutter_naver_map")
//    channel.setMethodCallHandler(FlutterNaverMapPlugin());
    pluginBinding = flutterPluginBinding
    val messenger = flutterPluginBinding.getBinaryMessenger()
    pluginBinding!!
            .getFlutterEngine()
            .getPlatformViewsController()
            .getRegistry()
            .registerViewFactory(
                    "flutter_naver_map", FlutterNaverMapViewFactory(messenger, /*containerView=*/ null))
  }

  // This static function is optional and equivalent to onAttachedToEngine. It supports the old
  // pre-Flutter-1.12 Android projects. You are encouraged to continue supporting
  // plugin registration via this function while apps migrate to use the new Android APIs
  // post-flutter-1.12 via https://flutter.dev/go/android-project-migration.
  //
  // It is encouraged to share logic between onAttachedToEngine and registerWith to keep
  // them functionally equivalent. Only one of onAttachedToEngine or registerWith will be called
  // depending on the user's project. onAttachedToEngine or registerWith must both be defined
  // in the same class.
  companion object {
    @JvmStatic
    fun registerWith(registrar: Registrar) {
//      val channel = MethodChannel(registrar.messenger(), "flutter_naver_map")
//      channel.setMethodCallHandler(FlutterNaverMapPlugin())
      registrar
              .platformViewRegistry()
              .registerViewFactory(
                      "flutter_naver_map",
                      FlutterNaverMapViewFactory(registrar.messenger(), registrar.view()))
    }
  }

  override fun onDetachedFromEngine(@NonNull binding: FlutterPlugin.FlutterPluginBinding) {
    pluginBinding = null
  }
}
